ENH: Implement correct scalar and integer overflow errors for NEP 50

seberg opened the pull request:
This adds the main missing chunk of NEP 50 support.  Unfortunately, I feel this is still fairly complex, although I am not sure it can be helped...

For the ufuncs, it may be possible to move the special handling to later.  For the scalars, the special handling has a bit of additional complexity because of the "defer" check (if we did it always it would be easier logic flow).
Unfortunately, for the scalars it makes sense to keep the "faster" version for the safe cast around (for speed reasons).

The other question is whether we want to do this as a distinct method on the DType, although I don't think we have to settle that yet.  It should be easy to change.

A small oddity is that the user has to `resolve` from one of the "int" dtypes (int32, int64, uint64, object).  That seems OK to me, although it also seems like something that we may well know better later.

---

Marking as draft, I need to let this sink in a bit to see if I have a better idea.  Bug, this should be ready for a test ride and may complete NEP 50 logic (with the exception of possibly allowing Python scalars for example in `np.can_cast(123, np.int8))`

seberg committed:
ENH: Implement safe integers and weak python scalars for scalars

This requires adding a path that uses the "normal" Python object
to dtype conversion (setitem) function, rather than always converting
to the default dtype.

seberg committed:
ENH: Get correct integer conversion into ufuncs

seberg committed:
TST: Fix int overflow test for windows platforms where message differs

seberg committed:
MAINT: Use Python integers for int to array of uint32 calculation

The new weak promotion preserves the original type, this makes the
`//= 2**32` fail for certain inputs.  The alternative would be
typing that as `np.int64(2**32)`, but using Python integers seems
easier and cleaner.

The code was effectively OK before, since the inputs were guaranteed
signed (or Python integers) at that point and 2**32 would have been
considered like a NumPy `int64`.  (Which would be an alternative fix.)

seberg committed:
MAINT: Make mem-overlap tests weak scalar promotion ready

seberg committed:
BUG: Ensure new-style promotion is not accidentally used for some ints

This is the *actual* correct fix for the test adaptations.  The test
adaptations should only be necessary when running in weak-promotion mode,
but they are NOT doing that currently.

seberg committed:
TST: Improve tests for python integer special cases

seberg commented:
OK, I still don't love the scalar code, but considering that the ufunc code is at least well localized and the scalar code could use more cleanup (as in conversion function introduction probably via C++), I will mark this as ready for review.  I simply did not have a clear idea for how to improve it meaningfully.

mattip commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -987,13 +1014,21 @@ convert_to_@name@(PyObject *value, @type@ *result, npy_bool *may_need_deferring)
+            if (!PyTypeNum_ISCOMPLEX(NPY_@TYPE@)) {
+                return PROMOTION_REQUIRED;
+            }
+            return CONVERT_PYSCALAR;
```
Is this line really skipped during testing or is codecoverage mistaken?

mattip commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -1247,6 +1293,16 @@ static PyObject *
             return PyGenericArrType_Type.tp_as_number->nb_@oper@(a,b);
+        case CONVERT_PYSCALAR:
+            if (@ONAME@_setitem(other, (char *)&other_val, NULL) < 0) {
+                return NULL;
```
Is codecoverage mistaken here? This seems like it should be covered

mattip commented:
> whether we want to do this as a distinct method on the DType

I would prefer not to, if we can avoid more methods

seberg committed:
TST: Rebustify NEP 50 integer tests for windows and 32bit linux

seberg committed:
TST: Avoid possible warning from unnecessary cast with uninitialized values

This should be fixed in `choose` to not do the unnecessary cast,
see gh-22237.

seberg committed:
TST: Add test to cover complex to complex64 downcast in weak promotion

seberg committed:
TST: Cover additional NEP 50 scalar paths

Especially adding coverage also for the power operator which
does not share its code perfectly.

seberg commented:
>> whether we want to do this as a distinct method on the DType

> I would prefer not to, if we can avoid more methods

Hmm, trying to untangle the state of this.  The reason for custom handling was that we can allow these to remain unmodified:
```
np.uint8(-1)
# And also:
np.array([-1], np.uint8)
arr_uint8[0] = -1
```
i.e. we still allow the forced cast in some cases, but not in the *new* paths.

This seems almost exclusively relevant for that use-case where you use -1 to mean `UINT_MAX`, but *is* a relevant use-case there.

However, we are already changing quite a lot, so maybe it is OK to just change that as well, with the one possible exception that we could hard-code `np.uint(-1)` to keep working (but _not_ the other two examples).

The *other* question is only relevant *if* we do not want to change the above.  That was where I was coming from.  But I don't really see that it makes things clearer by not being a method (since methods are the typical path now, I think).

seberg commented:
Lets discuss this on Wednesday a bit, maybe we can make a decision then (I think the code coverage/tests should be fine now, but will fix it up if not).

seberg committed:
DOC,TST: Improve overflow error message and adjust test

seberg commented:
OK, this is rebased to account for gh-22385.  (I will have another scroll through myself later today just to make sure there are no small cleanups needed still).

mattip commented on numpy/core/tests/test_nep50_promotions.py:
```diff
@@ -61,12 +64,116 @@ def test_nep50_examples():
+@pytest.mark.parametrize("dtype", np.typecodes["AllInteger"])
+def test_nep50_weak_integers(dtype):
+    # Avoids warning (different code path for scalars)
+    np._set_promotion_state("weak")
```
Should the tests reset global state at finish?

seberg replied:
Yes, there is a module wide auto-fixture that defaults to `"weak_and_warn"` and resets the state at the end.

seberg committed:
MAINT: Minor include fixup in scalarmath after rebase and comment fixup

seberg commented:
Can we move forward with this, soon?  I would like to make sure we have full optional NEP 50 logic by next release for testing.
I suppose this is the last very important thing, maybe `np.can_cast(123, dtype=np.uint8)` should also be "special" (but even if, I suspect it has much less impact in practice).

mattip commented:
Thanks @seberg
